File: notification_system_impl_python/providers/local_provider.py
import json
from datetime import datetime
from typing import Dict, Any
from providers.base_provider import NotificationProvider
from dotenv import load_dotenv

load_dotenv()


class LocalProvider(NotificationProvider):
    def provider_name(self) -> str:
        return "local"

    def send(self, notification) -> Dict[str, Any]:
        try:
            payload = json.loads(notification.payload)
            message_type = notification.message_type.value if hasattr(notification.message_type, 'value') else str(notification.message_type)
            print("\n" + "-" * 55)
            print(f"LOCAL NOTIFICATION - {message_type.upper()}")
            print("-" * 55)
            print(f"Notification ID: {notification.id}")
            print(f"User ID:         {notification.user_id}")
            print(f"Type:            {message_type}")
            print(f"Provider:        {notification.provider.value if hasattr(notification.provider, 'value') else str(notification.provider)}")
            print(f"Time:            {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
            print("-" * 55)
            for key, value in payload.items():
                if key == 'body' and len(str(value)) > 100:
                    print(f"{key.capitalize()}: {str(value)[:100]}...")
                else:
                    print(f"{key.capitalize()}: {value}")

            print("-" * 55 + "\n")

            return {
                'success': True,
                'message': 'Notification logged locally',
                'response': {'notification_id': notification.id, 'payload': payload}
            }
        except Exception as e:
            return {
                'success': False,
                'message': f'Local provider failed: {str(e)}'
            }

File: notification_system_impl_go/core/providers/local_provider.go
package providers

import (
	"context"
	"encoding/json"
	"fmt"
	"time"

	"notification_system/models"
)

type LocalProvider struct{}

func NewLocalProvider() *LocalProvider {
	return &LocalProvider{}
}

func (l *LocalProvider) ProviderName() string {
	return "local"
}

func (l *LocalProvider) Send(ctx context.Context, notification *models.Notification) error {
	var payload map[string]interface{}
	if err := json.Unmarshal([]byte(notification.Payload), &payload); err != nil {
		return fmt.Errorf("invalid payload: %w", err)
	}

	fmt.Println()
	fmt.Println("═══════════════════════════════════════════════════════")
	fmt.Printf("📧 LOCAL NOTIFICATION - %s\n", notification.MessageType)
	fmt.Println("═══════════════════════════════════════════════════════")
	fmt.Printf("Notification ID: %s\n", notification.ID)
	fmt.Printf("User ID:         %s\n", notification.UserID)
	fmt.Printf("Type:            %s\n", notification.MessageType)
	fmt.Printf("Provider:        %s\n", notification.Provider)
	fmt.Printf("Time:            %s\n", time.Now().Format("2006-01-02 15:04:05"))
	fmt.Println("-------------------------------------------------------")

	for key, value := range payload {
		if key == "body" {
			bodyStr := fmt.Sprintf("%v", value)
			if len(bodyStr) > 100 {
				fmt.Printf("%s: %s...\n", key, bodyStr[:100])
			} else {
				fmt.Printf("%s: %v\n", key, value)
			}
		} else {
			fmt.Printf("%s: %v\n", key, value)
		}
	}

	fmt.Println("═══════════════════════════════════════════════════════")
	fmt.Println()

	notification.Status = models.StatusSent
	now := time.Now().UnixMilli()
	notification.SentAt = &now
	response := "Notification logged locally"
	notification.ProviderResponse = &response

	return nil
}
